src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/SystemTextJsonLanguageServer.cs

```diff
@@ -0,0 +1,107 @@
+#nullable enable
+
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using StreamJsonRpc;
+
+namespace Microsoft.CommonLanguageServerProtocol.Framework;
+
+internal abstract class SystemTextJsonLanguageServer<TRequestContext>(JsonRpc jsonRpc, JsonSerializerOptions options, ILspLogger logger) : AbstractLanguageServer<TRequestContext>(jsonRpc, logger)
+{
+    /// <summary>
+    /// JsonSerializer options used by streamjsonrpc (and for serializing / deserializing the requests to streamjsonrpc).
+    /// These options are specifically from the <see cref="StreamJsonRpc.SystemTextJsonFormatter"/> that added the exotic type converters.
+    /// </summary>
+    private readonly JsonSerializerOptions _jsonSerializerOptions = options;
+    protected override DelegatingEntryPoint CreateDelegatingEntryPoint(string method, IGrouping<string, RequestHandlerMetadata> handlersForMethod)
+    {
+        return new SystemTextJsonDelegatingEntryPoint(method, handlersForMethod, this);
+    }
+
+    protected virtual string GetLanguageForRequest(string methodName, JsonElement? parameters)
+    {
+        Logger.LogInformation($"Using default language handler for {methodName}");
+        return LanguageServerConstants.DefaultLanguageName;
+    }
+
+    private sealed class SystemTextJsonDelegatingEntryPoint(
+        string method,
+        IGrouping<string, RequestHandlerMetadata> handlersForMethod,
+        SystemTextJsonLanguageServer<TRequestContext> target) : DelegatingEntryPoint(method, handlersForMethod)
+    {
+        private static readonly MethodInfo s_parameterlessEntryPoint = typeof(SystemTextJsonDelegatingEntryPoint).GetMethod(nameof(SystemTextJsonDelegatingEntryPoint.ExecuteRequest0Async), BindingFlags.NonPublic | BindingFlags.Instance)!;
+        private static readonly MethodInfo s_entryPoint = typeof(SystemTextJsonDelegatingEntryPoint).GetMethod(nameof(SystemTextJsonDelegatingEntryPoint.ExecuteRequestAsync), BindingFlags.NonPublic | BindingFlags.Instance)!;
+
+        public override MethodInfo GetEntryPoint(bool hasParameter)
+        {
+            return hasParameter ? s_entryPoint : s_parameterlessEntryPoint;
+        }
+
+        /// <summary>
+        /// StreamJsonRpc entry point for handlers with no parameters.
+        /// Unlike Newtonsoft, we have to differentiate instead of using default parameters.
+        /// </summary>
+        private Task<JsonElement?> ExecuteRequest0Async(CancellationToken cancellationToken = default)
+        {
+            return ExecuteRequestAsync(null, cancellationToken);
+        }
+
+        /// <summary>
+        /// StreamJsonRpc entry point for handlers with parameters (and any response) type.
+        /// </summary>
+        private async Task<JsonElement?> ExecuteRequestAsync(JsonElement? request, CancellationToken cancellationToken = default)
```

Comment:
My first approach was to attempt to generify / abstract out the serialization part into a very small piece and re-use the same entry point code, due to the different token types between newtonsoft and STJ.

STJ uses JsonElement, which is a struct, whereas newtonsoft uses JToken (a class).  This caused all sorts of nullability differences between the two impls, so 100x easier to separate them (even if there is a small amount of duplication).

And as mentioned, all of this is simplified once we switch Razor/XAML to our types and remove the custom serialization logic here.